File: backend/service/admin/biz/base_job.go
package biz

import (
	"context"

	adminv1 "shop/api/gen/go/admin/v1"
	"shop/pkg/biz"
	"shop/pkg/errorsx"
	"shop/pkg/gen/data"
	"shop/pkg/gen/models"
	"shop/pkg/job"

	_mapper "github.com/liujitcn/go-utils/mapper"
	_string "github.com/liujitcn/go-utils/string"
	"github.com/liujitcn/gorm-kit/repository"
)

// BaseJobCase 定时任务业务实例
type BaseJobCase struct {
	*biz.BaseCase
	*data.BaseJobRepository
	baseJobLogCase *BaseJobLogCase
	cronServer     *job.CronServer
	formMapper     *_mapper.CopierMapper[adminv1.BaseJobForm, models.BaseJob]
	mapper         *_mapper.CopierMapper[adminv1.BaseJob, models.BaseJob]
}

// NewBaseJobCase 创建定时任务业务实例
func NewBaseJobCase(baseCase *biz.BaseCase, baseJobRepo *data.BaseJobRepository, baseJobLogCase *BaseJobLogCase, cronServer *job.CronServer) *BaseJobCase {
	formMapper := _mapper.NewCopierMapper[adminv1.BaseJobForm, models.BaseJob]()
	formMapper.AppendConverters(_mapper.NewJSONTypeConverter[[]*adminv1.BaseJobArgs]().NewConverterPair())
	mapper := _mapper.NewCopierMapper[adminv1.BaseJob, models.BaseJob]()
	mapper.AppendConverters(_mapper.NewJSONTypeConverter[[]*adminv1.BaseJobArgs]().NewConverterPair())

	return &BaseJobCase{
		BaseCase:          baseCase,
		BaseJobRepository: baseJobRepo,
		baseJobLogCase:    baseJobLogCase,
		cronServer:        cronServer,
		formMapper:        formMapper,
		mapper:            mapper,
	}
}

// PageBaseJobs 分页查询定时任务
func (c *BaseJobCase) PageBaseJobs(ctx context.Context, req *adminv1.PageBaseJobsRequest) (*adminv1.PageBaseJobsResponse, error) {
	query := c.Query(ctx).BaseJob
	opts := make([]repository.QueryOption, 0, 4)
	opts = append(opts, repository.Order(query.CreatedAt.Desc()))
	// 传入任务名称时，按名称模糊匹配定时任务。
	if req.GetName() != "" {
		opts = append(opts, repository.Where(query.Name.Like("%"+req.GetName()+"%")))
	}
	// 传入调用目标时，按调用目标模糊匹配定时任务。
	if req.GetInvokeTarget() != "" {
		opts = append(opts, repository.Where(query.InvokeTarget.Like("%"+req.GetInvokeTarget()+"%")))
	}
	if req.Status != nil {
		opts = append(opts, repository.Where(query.Status.Eq(int32(req.GetStatus()))))
	}

	list, total, err := c.Page(ctx, req.GetPageNum(), req.GetPageSize(), opts...)
	if err != nil {
		return nil, err
	}

	resList := make([]*adminv1.BaseJob, 0, len(list))
	for _, item := range list {
		baseJob := c.mapper.ToDTO(item)
		resList = append(resList, baseJob)
	}
	return &adminv1.PageBaseJobsResponse{BaseJobs: resList, Total: int32(total)}, nil
}

// GetBaseJob 获取定时任务
func (c *BaseJobCase) GetBaseJob(ctx context.Context, id int64) (*adminv1.BaseJobForm, error) {
	baseJob, err := c.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	res := c.formMapper.ToDTO(baseJob)
	return res, nil
}

// CreateBaseJob 创建定时任务
func (c *BaseJobCase) CreateBaseJob(ctx context.Context, req *adminv1.BaseJobForm) error {
	baseJob := c.formMapper.ToEntity(req)
	baseJob.Args = _string.ConvertAnyToJsonString(req.GetArgs())
	err := c.Create(ctx, baseJob)
	if err != nil {
		// 命中调用目标唯一索引冲突时，返回稳定的业务冲突错误。
		if errorsx.IsMySQLDuplicateKey(err) {
			return errorsx.UniqueConflict("调用目标重复", "base_job", "invoke_target", "unique_base_job").WithCause(err)
		}
		return err
	}
	return nil
}

// UpdateBaseJob 更新定时任务
func (c *BaseJobCase) UpdateBaseJob(ctx context.Context, req *adminv1.BaseJobForm) error {
	baseJob := c.formMapper.ToEntity(req)
	baseJob.Args = _string.ConvertAnyToJsonString(req.GetArgs())
	err := c.UpdateByID(ctx, baseJob)
	if err != nil {
		// 命中调用目标唯一索引冲突时，返回稳定的业务冲突错误。
		if errorsx.IsMySQLDuplicateKey(err) {
			return errorsx.UniqueConflict("调用目标重复", "base_job", "invoke_target", "unique_base_job").WithCause(err)
		}
		return err
	}
	return nil
}

// DeleteBaseJob 删除定时任务
func (c *BaseJobCase) DeleteBaseJob(ctx context.Context, id string) error {
	return c.DeleteByIDs(ctx, _string.ConvertStringToInt64Array(id))
}

// SetBaseJobStatus 设置定时任务状态
func (c *BaseJobCase) SetBaseJobStatus(ctx context.Context, req *adminv1.SetBaseJobStatusRequest) error {
	return c.UpdateByID(ctx, &models.BaseJob{
		ID:     req.GetId(),
		Status: req.GetStatus(),
	})
}

// StartBaseJob 启动定时任务
func (c *BaseJobCase) StartBaseJob(ctx context.Context, req *adminv1.StartBaseJobRequest) error {
	baseJob, err := c.FindByID(ctx, req.GetId())
	if err != nil {
		return err
	}
	return c.cronServer.StartJob(ctx, baseJob)
}

// StopBaseJob 停止定时任务
func (c *BaseJobCase) StopBaseJob(ctx context.Context, req *adminv1.StopBaseJobRequest) error {
	baseJob, err := c.FindByID(ctx, req.GetId())
	if err != nil {
		return err
	}
	return c.cronServer.StopJob(ctx, baseJob)
}

// ExecuteBaseJob 立即执行定时任务
func (c *BaseJobCase) ExecuteBaseJob(ctx context.Context, req *adminv1.ExecuteBaseJobRequest) error {
	baseJob, err := c.FindByID(ctx, req.GetId())
	if err != nil {
		return err
	}
	return c.cronServer.RunJob(ctx, baseJob)
}

File: frontend/admin/src/api/admin/base_job.ts
import service from "@/utils/request";
import {
  type BaseJobForm,
  type BaseJobLog,
  type BaseJobService,
  type CreateBaseJobRequest,
  type DeleteBaseJobRequest,
  type ExecuteBaseJobRequest,
  type GetBaseJobLogRequest,
  type GetBaseJobRequest,
  type PageBaseJobLogsRequest,
  type PageBaseJobLogsResponse,
  type PageBaseJobsRequest,
  type PageBaseJobsResponse,
  type SetBaseJobStatusRequest,
  type StartBaseJobRequest,
  type StopBaseJobRequest,
  type UpdateBaseJobRequest
} from "@/rpc/admin/v1/base_job";
import type { Empty } from "@/rpc/google/protobuf/empty";

const BASE_JOB_URL = "/v1/admin/base/job";
const BASE_JOB_LOG_URL = "/v1/admin/base/job-log";

/** Admin定时任务服务 */
export class BaseJobServiceImpl implements BaseJobService {
  /** 查询定时任务分页列表 */
  PageBaseJobs(request: PageBaseJobsRequest): Promise<PageBaseJobsResponse> {
    return service<PageBaseJobsRequest, PageBaseJobsResponse>({
      url: `${BASE_JOB_URL}`,
      method: "get",
      params: request
    });
  }

  /** 查询定时任务 */
  GetBaseJob(request: GetBaseJobRequest): Promise<BaseJobForm> {
    return service<GetBaseJobRequest, BaseJobForm>({
      url: `${BASE_JOB_URL}/${request.id}`,
      method: "get"
    });
  }

  /** 创建定时任务 */
  CreateBaseJob(request: CreateBaseJobRequest): Promise<Empty> {
    return service<BaseJobForm | undefined, Empty>({
      url: `${BASE_JOB_URL}`,
      method: "post",
      data: request.base_job
    });
  }

  /** 更新定时任务 */
  UpdateBaseJob(request: UpdateBaseJobRequest): Promise<Empty> {
    return service<BaseJobForm | undefined, Empty>({
      url: `${BASE_JOB_URL}/${request.base_job?.id ?? ""}`,
      method: "put",
      data: request.base_job
    });
  }

  /** 删除定时任务 */
  DeleteBaseJob(request: DeleteBaseJobRequest): Promise<Empty> {
    return service<DeleteBaseJobRequest, Empty>({
      url: `${BASE_JOB_URL}/${request.id}`,
      method: "delete"
    });
  }

  /** 设置状态 */
  SetBaseJobStatus(request: SetBaseJobStatusRequest): Promise<Empty> {
    return service<SetBaseJobStatusRequest, Empty>({
      url: `${BASE_JOB_URL}/${request.id}/status`,
      method: "put",
      data: request
    });
  }

  /** 启动任务 */
  StartBaseJob(request: StartBaseJobRequest): Promise<Empty> {
    return service<StartBaseJobRequest, Empty>({
      url: `${BASE_JOB_URL}/${request.id}/running`,
      method: "put",
      data: request
    });
  }

  /** 停止任务 */
  StopBaseJob(request: StopBaseJobRequest): Promise<Empty> {
    return service<StopBaseJobRequest, Empty>({
      url: `${BASE_JOB_URL}/${request.id}/running`,
      method: "delete",
      data: request
    });
  }

  /** 执行任务 */
  ExecuteBaseJob(request: ExecuteBaseJobRequest): Promise<Empty> {
    return service<ExecuteBaseJobRequest, Empty>({
      url: `${BASE_JOB_URL}/${request.id}/execution`,
      method: "post",
      data: request
    });
  }

  /** 查询定时任务日志分页列表 */
  PageBaseJobLogs(request: PageBaseJobLogsRequest): Promise<PageBaseJobLogsResponse> {
    return service<PageBaseJobLogsRequest, PageBaseJobLogsResponse>({
      url: `${BASE_JOB_LOG_URL}`,
      method: "get",
      params: request
    });
  }

  /** 查询定时任务日志 */
  GetBaseJobLog(request: GetBaseJobLogRequest): Promise<BaseJobLog> {
    return service<GetBaseJobLogRequest, BaseJobLog>({
      url: `${BASE_JOB_LOG_URL}/${request.id}`,
      method: "get",
      params: request
    });
  }
}

export const defBaseJobService = new BaseJobServiceImpl();
